### PR TITLE
[Merged by Bors] - chore: more universe generalisations / fixes

### PR DESCRIPTION
**Pull request by kbuzzard**
There are changes of two types: first I add some `.{w}` in some declarations to ensure that universes are in the right order. Secondly I generalise some results from `Category.{max u1 v1, u2}` to `Category.{v2, u2}`.

From the Copenhagen workshop.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by kbuzzard** [comment]
Some results about functors C.{v1,u1} => D.{v2,u2} really do need an assumption of the form v2>=max(v1,u1). For example in the theory of Grothendieck topologies (D-valued presheaves on C) you want to take a product in D over some collection of arrows indexed by a collection of objects in C, and this is often solved via tricks like defining v2 to be max(v1,u1). However overuse of this ended up putting universe restrictions in random places. For example the type of `CategoryTheory.sheafCompose` in `master` begins
```
CategoryTheory.sheafCompose.{v₁, u₁, u₂, u₃} {C : Type u₁} [inst✝ : Category.{v₁, u₁} C] {A : Type u₂}
  [inst✝¹ : Category.{max v₁ u₁, u₂} A] {B : Type u₃} [inst✝² : Category.{max v₁ u₁, u₃} B]
...
```
and in particular the morphism universes for A and B had to be equal. In particular it was not possible to use `CategoryTheory.sheafCompose` to bump universe levels.

After this PR, the universes for `sheafCompose` are completely general:

```
CategoryTheory.sheafCompose.{v₁, v₂, v₃, u₁, u₂, u₃} {C : Type u₁} [inst✝ : Category.{v₁, u₁} C] {A : Type u₂}
  [inst✝¹ : Category.{v₂, u₂} A] {B : Type u₃} [inst✝² : Category.{v₃, u₃} B] (J : GrothendieckTopology.{v₁, u₁} C)
  (F : Functor.{v₂, v₃, u₂, u₃} A B)
```

but the results which actually need the extra hypothesis still have them; I call these universe-restricted variables `A'` and `P'` to distinguish them from the universe polymorphic `A` and `P`.

File: Mathlib/CategoryTheory/Sites/Sheaf.lean [hunk 1]
```diff
@@ -457,12 +457,16 @@ namespace Presheaf
 -- between 00VQ and 00VR.
 variable {C : Type u₁} [Category.{v₁} C]
 
-variable {A : Type u₂} [Category.{max v₁ u₁} A]
+variable {A' : Type u₂} [Category.{max v₁ u₁} A'] -- to use if we can't get `A` working
+
+variable {A : Type u₂} [Category.{v₂} A]
```
Comment (ocfnash) [inline Mathlib/CategoryTheory/Sites/Sheaf.lean:462]: I think it's easier to follow the design if we declare `A` first. 
```suggestion
variable {A : Type u₂} [Category.{v₂} A]
variable {A' : Type u₂} [Category.{max v₁ u₁} A'] -- An alternative to `A` with universes satisfying inequalities.
```

File: Mathlib/CategoryTheory/Sites/Sheaf.lean [hunk 2]
```diff
@@ -457,12 +457,16 @@ namespace Presheaf
...
 
 variable (J : GrothendieckTopology C)
 
 variable {U : C} (R : Presieve U)
 
+variable (P' : Cᵒᵖ ⥤ A')
+
 variable (P : Cᵒᵖ ⥤ A)
```
Comment (ocfnash) [inline Mathlib/CategoryTheory/Sites/Sheaf.lean:470]:
```suggestion
variable (P : Cᵒᵖ ⥤ A) (P' : Cᵒᵖ ⥤ A')
```

File: Mathlib/CategoryTheory/Sites/Sheaf.lean [hunk 3]
```diff
@@ -583,6 +588,7 @@ def IsSheaf' (P : Cᵒᵖ ⥤ A) : Prop :=
   ∀ (U : C) (R : Presieve U) (_ : generate R ∈ J U), Nonempty (IsLimit (Fork.ofι _ (w R P)))
 #align category_theory.presheaf.is_sheaf' CategoryTheory.Presheaf.IsSheaf'
 
+-- open question: can we make type of `s` any more general?
```
Comment (ocfnash) [inline Mathlib/CategoryTheory/Sites/Sheaf.lean:591]:
```suggestion
-- TODO: Can the universes in `s` be generalised?
```

**Review by ocfnash — APPROVED** [review]
I note that [UnivLE](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Logic/UnivLE.html#UnivLE) was created after the work here was done, so I think it would be worth exploring if it could be used instead of the `A'` and `P'` trick proposed.

However, this is clearly a strict improvement either way.

Thanks!

bors d+

**Comment by kbuzzard** [comment]
I tried to use UnivLE but couldn't get it to work. I asked about it [here](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/UnivLE.20in.20CategoryTheory.2ESites.2ESheafOfTypes.20.3F/near/380295399).

bors r+

**Pull request closed without merging**